lib/cog/command/pipeline/executor.ex

```diff
@@ -667,27 +671,35 @@ defmodule Cog.Command.Pipeline.Executor do
   end
 
   defp assign_relay(relays, bundle_name, bundle_version) do
-    case Map.get(relays, bundle_name) do
-      # No assignment so let's pick one
-      nil ->
-        case Relays.pick_one(bundle_name, bundle_version) do
-          # No relays available
-          nil ->
-            {nil, relays}
-          # Store the selected relay in the relay cache
-          relay ->
-            {relay, Map.put(relays, bundle_name, relay)}
-        end
-      # Relay was previously assigned
-      relay ->
-        # Is the bundle still available on the relay? If not, remove the current assignment from the cache
-        # and select a new relay
-        if Relays.relay_available?(relay, bundle_name, bundle_version) do
-          {relay, relays}
-        else
-          relays = Map.delete(relays, bundle_name)
-          assign_relay(relays, bundle_name, bundle_version)
-        end
+    # If the bundle is assigned to a relay group carry on, otherwise bail
```

Comment:
Flipping the logic will reduce the number of DB queries issued to execute a pipeline:

```
case Map.get(relays, bundle_name) do
  # No assignment so let's pick one
  nil ->
    case Relays.pick_one(bundle_name, bundle_version) do
      # No relays available
      {:error, :no_relays} ->
        {nil, relays}
        # relays exist, but are disabled
      {:error, :no_enabled_relays} ->
        # Query database to clarify error condition before reporting to user
        if Cog.Repository.Bundles.assigned_to_group?(bundle_name) do
          {:no_enabled_relays, relays}
        else
          {:no_relay_group, relays}
        end
      # Store the selected relay in the relay cache
      {:ok, relay} ->
        {relay, Map.put(relays, bundle_name, relay)}
    end
    # Relay was previously assigned
  relay ->
  # Is the bundle still available on the relay? If not, remove the current assignment from the cache
  # and select a new relay
    if Relays.relay_available?(relay, bundle_name, bundle_version) do
      {relay, relays}
    else
      relays = Map.delete(relays, bundle_name)
      assign_relay(relays, bundle_name, bundle_version)
    end
end
```